autowarefoundation/autoware.universe
# refactor(autoware_tensorrt_common): multi-TensorRT compatibility & tensorrt_common as unified lib for all perception components

## Description

Incoming TensorRT upgrade requires refactor within perception components. The new version of `autoware_tensorrt_common` library will stand as high-level TensorRT API. PR considers compatibility with current environment (TensorRT 8.6) and future upgrade (TensorRT 10.7+).

This PR should be merged before TensorRT upgrade.

## Related links

**Parent Issue:**

- https://github.com/autowarefoundation/autoware/issues/5607



## How was this PR tested?

A. Current environment (without dependencies upgrade).
B. New environment (with dependencies upgrade).
* If rosdistro PR not merged (see parent issue for more information), clone [tensorrt_cmake_module](https://github.com/tier4/tensorrt_cmake_module) into e.g. `src/universe/external`.
```shell
git clone https://github.com/tier4/tensorrt_cmake_module.git src/universe/external/tensorrt_cmake_module
```
* Checkout for Autoware environment upgrade branch.
```shell
git fetch origin pull/5608/head:feat/cuda-trt-upgrade
git switch feat/cuda-trt-upgrade
```
* Remove [these lines](https://github.com/autowarefoundation/autoware/blob/c490050599aef24e7d82e5db33d920ac58e93495/docker/build.sh#L108-L109) to make sure build script will not change PR branch.
* Build docker locally.
```shell
./docker/build.sh --devel-only
```
* Test this PR in docker.

C. Perception evaluator [TIER IV internal link](https://evaluation.tier4.jp/evaluation/reports/487aa3eb-5568-570e-b6bc-2bba9fad9040?project_id=prd_jt)

## Notes for reviewers

* Implicit quantization Is deprecated and will be no longer supported in future. Here, this feature was delegated to another class definition which inherits base class of `TrtCommon`. [Documentation](https://docs.nvidia.com/deeplearning/tensorrt/release-notes/index.html#rel-10-6-0) says, _They may not give the optimal performance and accuracy. As a workaround, use INT8 explicit quantization instead_, therefore we might need to consider explicit quantization (during model deployment).
* You can test this PR in current environment. However, I encourage to test it also with upgraded libraries.
* Due to API changes, we can't split this PR.

## Interface changes

None.



## Effects on system behavior

None.
